File: basic_dsa_four_hundred_thirty_three.java
/**
 * Basic DSA Problems - Minimum Genetic Mutation (LeetCode 433)
 * Min mutations from start to end; each step = 1 char change, must be in bank
 */
public class basic_dsa_four_hundred_thirty_three {

    // Problem: 8-char gene A/C/G/T. One mutation = one char change. BFS.

    /**
     * BFS: for each gene, try changing each pos to A/C/G/T.
     * Time O(8 * 4 * n) = O(n), Space O(n)
     */
    public static int minMutation(String start, String end, String[] bank) {
        java.util.Set<String> set = new java.util.HashSet<>(java.util.Arrays.asList(bank));
        if (!set.contains(end)) return -1;
        char[] genes = {'A','C','G','T'};
        java.util.Queue<String> q = new java.util.LinkedList<>();
        java.util.Set<String> vis = new java.util.HashSet<>();
        q.offer(start); vis.add(start);
        int steps = 0;
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                String s = q.poll();
                if (s.equals(end)) return steps;
                char[] arr = s.toCharArray();
                for (int j = 0; j < 8; j++) {
                    char old = arr[j];
                    for (char g : genes) {
                        if (g == old) continue;
                        arr[j] = g;
                        String next = new String(arr);
                        if (set.contains(next) && !vis.contains(next)) {
                            vis.add(next);
                            q.offer(next);
                        }
                    }
                    arr[j] = old;
                }
            }
            steps++;
        }
        return -1;
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Minimum Genetic Mutation");
        System.out.println("=".repeat(60));

        System.out.println("AACCGGTT->AACCGGTA, [AACCGGTA] -> " + minMutation("AACCGGTT", "AACCGGTA", new String[]{"AACCGGTA"}));
        System.out.println("AACCGGTT->AAACGGTA, bank -> " + minMutation("AACCGGTT", "AAACGGTA", new String[]{"AACCGGTA", "AACCGCTA", "AAACGGTA"}));
        System.out.println("AAAAACCC->AACCCCCC, bank -> " + minMutation("AAAAACCC", "AACCCCCC", new String[]{"AAAACCCC", "AAACCCCC", "AACCCCCC"}));
    }
}

File: basic_dsa_four_hundred_thirty_three.cpp
/**
 * Basic DSA - Average Salary Excluding the Minimum and Maximum Salary (LeetCode 1491)
 * At least three distinct salaries; drop one min and one max, average the rest.
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <vector>

double average(const std::vector<int>& salary) {
    int mn = salary[0];
    int mx = salary[0];
    long long sum = 0;
    for (int x : salary) {
        mn = std::min(mn, x);
        mx = std::max(mx, x);
        sum += x;
    }
    return static_cast<double>(sum - mn - mx) / static_cast<double>(salary.size() - 2);
}

int main() {
    std::cout << std::string(60, '=') << "\n";
    std::cout << "Problem: Average Salary Excluding the Minimum and Maximum Salary\n";
    std::cout << std::string(60, '=') << "\n";

    std::vector<int> v1 = {4000, 3000, 1000, 2000};
    std::cout << "[4000,3000,1000,2000] -> " << average(v1) << "\n";

    std::vector<int> v2 = {1000, 2000, 3000};
    std::cout << "[1000,2000,3000] -> " << average(v2) << "\n";
}
